File: backend/app/api/schemas/evidence.py
from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field
from enum import Enum


class EvidenceType(str, Enum):
    ANOMALY_EVIDENCE = "ANOMALY_EVIDENCE"
    PATTERN_EVIDENCE = "PATTERN_EVIDENCE"
    CORRELATION_EVIDENCE = "CORRELATION_EVIDENCE"
    TEMPORAL_EVIDENCE = "TEMPORAL_EVIDENCE"
    TREND_EVIDENCE = "TREND_EVIDENCE"
    GROUP_DIFFERENCE_EVIDENCE = "GROUP_DIFFERENCE_EVIDENCE"
    CHANGE_POINT_EVIDENCE = "CHANGE_POINT_EVIDENCE"
    STATISTICAL_EVIDENCE = "STATISTICAL_EVIDENCE"


class EvidencePolarity(str, Enum):
    SUPPORT = "support"
    CONTRADICT = "contradict"
    NEUTRAL = "neutral"


class Evidence(BaseModel):
    evidence_id: str
    dataset_id: str
    title: str
    description: str
    evidence_type: str  # ANOMALY_EVIDENCE, PATTERN_EVIDENCE, etc.
    strength: int  # 0 - 100
    polarity: str = "support"  # support, contradict, neutral
    supports_pattern_ids: List[str] = Field(default_factory=list)
    related_finding_ids: List[str] = Field(default_factory=list)
    related_relationship_ids: List[str] = Field(default_factory=list)
    related_event_ids: List[str] = Field(default_factory=list)
    columns: List[str] = Field(default_factory=list)
    source_metadata: Dict[str, Any] = Field(default_factory=dict)


class EvidenceCluster(BaseModel):
    cluster_id: str
    dataset_id: str
    title: str
    summary: str
    strength: int  # 0 - 100
    evidence_ids: List[str] = Field(default_factory=list)
    finding_ids: List[str] = Field(default_factory=list)
    pattern_ids: List[str] = Field(default_factory=list)
    relationship_ids: List[str] = Field(default_factory=list)
    timeline_event_ids: List[str] = Field(default_factory=list)
    primary_columns: List[str] = Field(default_factory=list)


class Hypothesis(BaseModel):
    hypothesis_id: str
    dataset_id: str
    title: str
    statement: str
    status: str = "candidate"  # candidate, under_review
    confidence: int  # 0 - 100
    evidence_strength: int  # 0 - 100
    supporting_evidence_ids: List[str] = Field(default_factory=list)
    contradicting_evidence_ids: List[str] = Field(default_factory=list)
    supporting_pattern_ids: List[str] = Field(default_factory=list)
    related_relationship_ids: List[str] = Field(default_factory=list)
    timeline_event_ids: List[str] = Field(default_factory=list)
    reasoning: str
    primary_columns: List[str] = Field(default_factory=list)


class InvestigationThread(BaseModel):
    thread_id: str
    dataset_id: str
    title: str
    priority: int  # 0 - 100
    status: str = "open"  # open, active, resolved
    summary: str
    hypothesis_ids: List[str] = Field(default_factory=list)
    evidence_ids: List[str] = Field(default_factory=list)
    pattern_ids: List[str] = Field(default_factory=list)
    finding_ids: List[str] = Field(default_factory=list)
    timeline_event_ids: List[str] = Field(default_factory=list)
    primary_columns: List[str] = Field(default_factory=list)


class InvestigationGraphNode(BaseModel):
    id: str
    label: str
    type: str  # finding, pattern, evidence, hypothesis, event
    data: Dict[str, Any] = Field(default_factory=dict)


class InvestigationGraphEdge(BaseModel):
    id: str
    source: str
    target: str
    label: str  # SUPPORTS, CONTRADICTS, DERIVED_FROM, RELATED_TO, OCCURS_BEFORE
    strength: Optional[int] = None


class InvestigationGraphData(BaseModel):
    nodes: List[InvestigationGraphNode] = Field(default_factory=list)
    edges: List[InvestigationGraphEdge] = Field(default_factory=list)


class EvidenceSummary(BaseModel):
    dataset_id: str
    status: str  # completed, empty, failed
    total_evidence: int
    strong_evidence: int
    moderate_evidence: int
    weak_evidence: int
    total_hypotheses: int
    total_investigation_threads: int
    top_threads: List[InvestigationThread] = Field(default_factory=list)
    evidence: List[Evidence] = Field(default_factory=list)
    hypotheses: List[Hypothesis] = Field(default_factory=list)
    clusters: List[EvidenceCluster] = Field(default_factory=list)
    threads: List[InvestigationThread] = Field(default_factory=list)
    graph_data: InvestigationGraphData = Field(default_factory=InvestigationGraphData)
    analyzed_at: str
    duration_ms: Optional[float] = None


class EvidenceAnalyzeRequest(BaseModel):
    force: bool = False

File: frontend/src/types/evidence.ts
export type EvidenceType =
  | "ANOMALY_EVIDENCE"
  | "PATTERN_EVIDENCE"
  | "CORRELATION_EVIDENCE"
  | "TEMPORAL_EVIDENCE"
  | "TREND_EVIDENCE"
  | "GROUP_DIFFERENCE_EVIDENCE"
  | "CHANGE_POINT_EVIDENCE"
  | "STATISTICAL_EVIDENCE"
  | string;

export type EvidencePolarity = "support" | "contradict" | "neutral" | string;

export interface Evidence {
  evidence_id: string;
  dataset_id: string;
  title: string;
  description: string;
  evidence_type: EvidenceType;
  strength: number; // 0 - 100
  polarity: EvidencePolarity; // support, contradict, neutral
  supports_pattern_ids?: string[];
  related_finding_ids?: string[];
  related_relationship_ids?: string[];
  related_event_ids?: string[];
  columns: string[];
  source_metadata?: Record<string, any>;
}

export interface EvidenceCluster {
  cluster_id: string;
  dataset_id: string;
  title: string;
  summary: string;
  strength: number; // 0 - 100
  evidence_ids: string[];
  finding_ids: string[];
  pattern_ids: string[];
  relationship_ids: string[];
  timeline_event_ids: string[];
  primary_columns: string[];
}

export interface Hypothesis {
  hypothesis_id: string;
  dataset_id: string;
  title: string;
  statement: string;
  status: string; // candidate, under_review, etc.
  confidence: number; // 0 - 100
  evidence_strength: number; // 0 - 100
  supporting_evidence_ids: string[];
  contradicting_evidence_ids: string[];
  contradictory_evidence_ids?: string[];
  supporting_pattern_ids: string[];
  related_relationship_ids?: string[];
  timeline_event_ids?: string[];
  reasoning: string;
  primary_columns: string[];
}

export interface InvestigationThread {
  thread_id: string;
  dataset_id: string;
  title: string;
  priority: number; // 0 - 100
  status: string; // open, active, resolved
  summary: string;
  hypothesis_ids: string[];
  evidence_ids: string[];
  pattern_ids: string[];
  finding_ids: string[];
  timeline_event_ids: string[];
  primary_columns: string[];
}

export interface InvestigationGraphNode {
  id: string;
  label: string;
  type: "finding" | "pattern" | "evidence" | "hypothesis" | "event" | string;
  data?: Record<string, any>;
}

export interface InvestigationGraphEdge {
  id: string;
  source: string;
  target: string;
  label: string;
  strength?: number | null;
}

export interface InvestigationGraphData {
  nodes: InvestigationGraphNode[];
  edges: InvestigationGraphEdge[];
}

export interface EvidenceSummary {
  dataset_id: string;
  status: "completed" | "empty" | "failed" | string;
  total_evidence: number;
  strong_evidence: number;
  moderate_evidence: number;
  weak_evidence: number;
  total_hypotheses: number;
  total_investigation_threads: number;
  top_threads: InvestigationThread[];
  evidence: Evidence[];
  hypotheses: Hypothesis[];
  clusters: EvidenceCluster[];
  threads: InvestigationThread[];
  graph_data: InvestigationGraphData;
  analyzed_at: string;
  duration_ms?: number | null;
}

export interface EvidenceAnalyzeRequest {
  force?: boolean;
}
